### PR TITLE
P2P API

### PR DESCRIPTION
**Pull request by bmeike**
this is PR https://github.com/couchbaselabs/couchbase-lite-java-ee-root/pull/6
but working...

Implementation of the P2P API.

### CODE REVIEW

File: common/main/java/com/couchbase/lite/AbstractReplicatorConfiguration.java [hunk 1]
```diff
@@ -261,13 +275,27 @@ public final ReplicatorConfiguration setPushFilter(ReplicationFilter pushFilter)
...
+
+    /**
+     * Sets the replicator verification mode.
+     * The default value is ServerCertificateVerificationMode.CA_CERT.
+     *
+     * @param mode Specifies the way the replicator verifies the server identity when using TLS communication
+     * @return this.
+     */
+    @NonNull
+    public final ReplicatorConfiguration setServerCertificateVerficationMode(
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/AbstractReplicatorConfiguration.java:295]: Typos : *Verfication

File: common/main/java/com/couchbase/lite/AbstractReplicatorConfiguration.java [hunk 2]
```diff
@@ -349,7 +385,15 @@ public final ReplicatorConfiguration setReplicatorType(@NonNull ReplicatorType r
     @NonNull
     public final Endpoint getTarget() { return target; }
 
+    /**
+     * Return the replicator verification mode.
+     */
+    @NonNull
+    public final ServerCertificateVerificationMode getServerCertificateVerficationMode() {
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/AbstractReplicatorConfiguration.java:392]: Typos: *Verfication.

File: common/main/java/com/couchbase/lite/BasicAuthenticator.java [hunk 3]
```diff
@@ -34,19 +36,37 @@
...
-        this.username = username;
-        this.password = password;
+    /**
+     * Create a Basic Authenticator.
+     * The new instance contains a copy of the password byte array parameter:
+     * the owner of the original retains the responsiblity for zeroing it before releasing it.
+     *
+     * @return
+     */
+    public BasicAuthenticator(@NonNull String username, @NonNull byte[] password) {
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/BasicAuthenticator.java:65]: I noticed that Java's KeyStore use char[] as the parameter type for password. Do you think we should do the same?   The spec said byte[] but if char[] is preferred, we can do char[] as well.

File: common/main/java/com/couchbase/lite/ListenerPasswordAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+package com.couchbase.lite;
+
+import android.support.annotation.Nullable;
+
+
+/**
+ * A Listener Password Authenticator
+ */
+public class ListenerPasswordAuthenticator implements ListenerAuthenticator {
+    public ListenerPasswordAuthenticator(@Nullable ListenerPasswordAuthenticatorDelegate delegate) { }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/ListenerPasswordAuthenticator.java:25]: The delegate is nonnull.

File: common/main/java/com/couchbase/lite/ListenerPasswordAuthenticatorDelegate.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+//
+package com.couchbase.lite;
+
+
+import android.support.annotation.NonNull;
+
+
+@FunctionalInterface
+public interface ListenerPasswordAuthenticatorDelegate {
+    boolean authenticate(@NonNull String username, @NonNull byte[] password);
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/ListenerPasswordAuthenticatorDelegate.java:24]: Same as my comment above regarding char[] vs byte[].

File: common/main/java/com/couchbase/lite/URLEndpointListener.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.net.URL;
+import java.util.List;
+
+
+public class URLEndpointListener {
+    private final URLEndpointListenerConfiguration config;
+
+    @Nullable
+    private int port;
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/URLEndpointListener.java:29]: I think int is not the Object so Nullable is not applicable here?

File: common/main/java/com/couchbase/lite/URLEndpointListener.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Nullable
+    public List<URL> getUrls() { return urls; }
+
+    public void setUrls(@Nullable List<URL> urls) { this.urls = urls; }
+
+    @Nullable
+    public ConnectionStatus getStatus() { return status; }
+
+    public void setStatus(@Nullable ConnectionStatus status) { this.status = status; }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/URLEndpointListener.java:54]: status is readonly.

File: common/main/java/com/couchbase/lite/URLEndpointListener.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    public URLEndpointListenerConfiguration getConfig() { return config; }
+
+    public int getPort() { return port; }
+
+    public void setPort(int port) { this.port = port; }
+
+    @Nullable
+    public List<URL> getUrls() { return urls; }
+
+    public void setUrls(@Nullable List<URL> urls) { this.urls = urls; }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/URLEndpointListener.java:49]: urls is readonly.

File: common/main/java/com/couchbase/lite/URLEndpointListener.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    private ConnectionStatus status;
+
+    // Constructors
+    public URLEndpointListener(@NonNull URLEndpointListenerConfiguration config) { this.config = config; }
+
+    public URLEndpointListenerConfiguration getConfig() { return config; }
+
+    public int getPort() { return port; }
+
+    public void setPort(int port) { this.port = port; }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/URLEndpointListener.java:44]: port is readonly.

File: common/main/java/com/couchbase/lite/URLEndpointListener.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+    public void setUrls(@Nullable List<URL> urls) { this.urls = urls; }
+
+    @Nullable
+    public ConnectionStatus getStatus() { return status; }
+
+    public void setStatus(@Nullable ConnectionStatus status) { this.status = status; }
+
+    // Methods
+    public void start() { }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/URLEndpointListener.java:57]: Missing throws CouchbaseLiteException.

File: common/main/java/com/couchbase/lite/internal/AbstractTLSIdentity.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    public static final String CERT_ATTR_REGISTERED_ID = "registeredID";
+
+    @Nullable
+    private List<Certificate> certs;
+
+    @Nullable
+    private Date expiration;
+
+    @Nullable
+    protected List<Certificate> getCerts() { return certs; }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/internal/AbstractTLSIdentity.java:50]: Should it be public?

File: common/main/java/com/couchbase/lite/internal/AbstractTLSIdentity.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    protected void setCerts(@Nullable List<Certificate> certs) { this.certs = certs; }
+
+    @Nullable
+    protected Date getExpiration() { return expiration; }
+
+    protected void setExpiration(@Nullable Date expiration) { this.expiration = expiration; }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/internal/AbstractTLSIdentity.java:57]: expiration is readonly.

File: common/main/java/com/couchbase/lite/internal/AbstractTLSIdentity.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    @Nullable
+    private List<Certificate> certs;
+
+    @Nullable
+    private Date expiration;
+
+    @Nullable
+    protected List<Certificate> getCerts() { return certs; }
+
+    protected void setCerts(@Nullable List<Certificate> certs) { this.certs = certs; }
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/internal/AbstractTLSIdentity.java:52]: certs is readonly.

File: java/main/java/com/couchbase/lite/TLSIdentity.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.Date;
+import java.util.Map;
+
+import com.couchbase.lite.internal.AbstractTLSIdentity;
+
+
+public class TLSIdentity extends AbstractTLSIdentity {
+
+    @Nullable
+    public TLSIdentity getIdentity(KeyStore keyStore, String alias, byte[] keyPassword) {
```
Comment (pasin) [inline java/main/java/com/couchbase/lite/TLSIdentity.java:30]: Missing throws CouchbaseLiteException.

File: java/main/java/com/couchbase/lite/TLSIdentity.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        return null;
+    }
+
+    public TLSIdentity createIdentity(
```
Comment (pasin) [inline java/main/java/com/couchbase/lite/TLSIdentity.java:34]: Missing throws CouchbaseLiteException.

File: java/main/java/com/couchbase/lite/TLSIdentity.java (hunk 14 above)
Comment (pasin) [reply java/main/java/com/couchbase/lite/TLSIdentity.java:30]: keyPassword is nullable.

File: android/main/java/com/couchbase/lite/TLSIdentity.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.List;
+import java.util.Map;
+
+import com.couchbase.lite.internal.AbstractTLSIdentity;
+
+
+public class TLSIdentity extends AbstractTLSIdentity {
+
+    @Nullable
+    public TLSIdentity getIdentity(@NonNull String alias, @Nullable byte[] keyPassword) {
```
Comment (pasin) [inline android/main/java/com/couchbase/lite/TLSIdentity.java:34]: Should be static method.

File: android/main/java/com/couchbase/lite/TLSIdentity.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+    //@NonNull
+    public TLSIdentity getIdentity(
+        @NonNull KeyStore.PrivateKeyEntry privateKey,
+        @NonNull List<Certificate> certificate) {
+        return null;
+    }
+
+    //@NonNull
+    public TLSIdentity createIdentity(
```
[the remark this answers is not in the document]
Comment (pasin) [reply android/main/java/com/couchbase/lite/TLSIdentity.java:46]: Should be static method.

File: android/main/java/com/couchbase/lite/TLSIdentity.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+    public TLSIdentity importIdentity(
+        @NonNull InputStream data,
+        @NonNull String dataType,
+        @Nullable byte[] dataPassword,
+        @NonNull String alias,
+        @Nullable byte[] keyPassword) {
+        return null;
+    }
+
+    public void deleteIdentity(@NonNull String alias) { }
```
[the remark this answers is not in the document]
Comment (pasin) [reply android/main/java/com/couchbase/lite/TLSIdentity.java:65]: Should be static method.

File: android/main/java/com/couchbase/lite/TLSIdentity.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+        boolean isServer,
+        @NonNull Map<String, String> attributes,
+        @Nullable Date expiration,
+        @NonNull String alias,
+        @Nullable byte[] keyPassword) {
+        return null;
+    }
+
+    //@NonNull
+    public TLSIdentity importIdentity(
```
[the remark this answers is not in the document]
Comment (pasin) [reply android/main/java/com/couchbase/lite/TLSIdentity.java:56]: Should be static method.

File: java/main/java/com/couchbase/lite/TLSIdentity.java (hunk 14 above)
Comment (pasin) [reply java/main/java/com/couchbase/lite/TLSIdentity.java:30]: Should be static method.

File: java/main/java/com/couchbase/lite/TLSIdentity.java (hunk 15 above)
Comment (pasin) [reply java/main/java/com/couchbase/lite/TLSIdentity.java:34]: Should be static method.

**Comment by bmeike** [comment]
All errors repaired.

**Pull request merged**